### PR TITLE
Enhance cloudprofileconfig for multi arch machine and filter machines ami based on arch

### PR DESCRIPTION
**Pull request by acumino**
**How to categorize this PR?**

/area delivery
/kind enhancement
/platform aws

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/backlog/issues/19

**Special notes for your reviewer**:
~~This PR is just the basic functionality implementation for the supporting arm64 machine. It will be made ready once gardener/gardener is released with the following PRs.~~

~~https://github.com/gardener/gardener/pull/6178~~
~~https://github.com/gardener/gardener/pull/6156~~

**Release note**:

```feature operator
`CloudProfileConfig` now supports a new field `.machineImages[].versions[].regions[].architecture`. It specifies the supported CPU architecture of the given machine image AMI.
```
```feature operator
`WorkerStatus` now supports a new field `.machineImage[].architecture`. It specifies the supported CPU architecture of the given worker pool.
```

### CODE REVIEW

File: pkg/apis/aws/v1alpha1/types_worker.go [hunk 1]
```diff
@@ -107,6 +107,9 @@ type MachineImage struct {
 	Version string `json:"version"`
 	// AMI is the AMI for the machine image.
 	AMI string `json:"ami"`
+	// Architecture is the CPU architecture of the machine image.
+	// +optional
+	Architecture *string `json:"architecture"`
```
Comment (rfranzke) [inline pkg/apis/aws/v1alpha1/types_worker.go:112]: `omitempty`

File: pkg/apis/aws/v1alpha1/types_cloudprofile.go [hunk 2]
```diff
@@ -52,4 +52,7 @@ type RegionAMIMapping struct {
 	Name string `json:"name"`
 	// AMI is the AMI for the machine image.
 	AMI string `json:"ami"`
+	// Architecture is the CPU architecture of the machine image.
+	// +optional
+	Architecture *string `json:"architecture"`
```
Comment (rfranzke) [inline pkg/apis/aws/v1alpha1/types_cloudprofile.go:57]: Please extend https://github.com/gardener/gardener-extension-provider-aws/blob/master/docs/usage-as-operator.md

File: pkg/apis/aws/helper/helper.go [hunk 3]
```diff
@@ -81,21 +81,21 @@ func FindSubnetForPurposeAndZone(subnets []api.Subnet, purpose, zone string) (*a
...
 
 // FindMachineImage takes a list of machine images and tries to find the first entry
-// whose name, version, and zone matches with the given name, version, and region. If no such entry is
+// whose name, version, architecture and zone matches with the given name, version, architecture and region. If no such entry is
 // found then an error will be returned.
-func FindMachineImage(machineImages []api.MachineImage, name, version string) (*api.MachineImage, error) {
+func FindMachineImage(machineImages []api.MachineImage, name, version string, arch *string) (*api.MachineImage, error) {
 	for _, machineImage := range machineImages {
-		if machineImage.Name == name && machineImage.Version == version {
+		if machineImage.Name == name && machineImage.Version == version && *machineImage.Architecture == *arch {
```
Comment (rfranzke) [inline pkg/apis/aws/helper/helper.go:88]: Use `pointer.StringEqual`?

Comment (rfranzke) [inline pkg/apis/aws/helper/helper.go:88]: What if `arch` is `nil`? Some folks might not have updated `gardener/gardener` when updating to this AWS extension version, so defaulting in `CloudProfile` or `Shoot` might not yet be in place.

File: pkg/apis/aws/helper/helper.go [hunk 4]
```diff
@@ -106,15 +106,15 @@ func FindAMIForRegionFromCloudProfile(cloudProfileConfig *api.CloudProfileConfig
 					continue
 				}
 				for _, mapping := range version.Regions {
-					if regionName == mapping.Name {
+					if regionName == mapping.Name && *arch == *mapping.Architecture {
```
Comment (rfranzke) [inline pkg/apis/aws/helper/helper.go:109]: same here

File: pkg/apis/aws/helper/helper.go (hunk 3 above)
Comment (acumino) [reply pkg/apis/aws/helper/helper.go:88]: Thanks, I didn't think of this scenario.

File: docs/usage-as-operator.md [hunk 5]
```diff
@@ -27,6 +27,7 @@ machineImages:
     regions:
     - name: eu-central-1
       ami: ami-034fd8c3f4026eb39
+      # architecture: amd64 # optional
```
Comment (rfranzke) [inline docs/usage-as-operator.md:30]: Maybe add a short sentence about this field after

> You have to map every version that you specify in `.spec.machineImages[].versions` here such that the AWS extension knows the AMI for every version you want to offer.

File: pkg/controller/worker/machine_images.go [hunk 6]
```diff
@@ -61,13 +61,21 @@ func (w *workerDelegate) findMachineImage(name, version string, region string, a
...
 
 		return machineImage.AMI, nil
 	}
 
-	return "", worker.ErrorMachineImageNotFound(name, version, *arch, region)
+	if arch != nil {
+		return "", worker.ErrorMachineImageNotFound(name, version, *arch, region)
+	} else {
+		return "", worker.ErrorMachineImageNotFound(name, version, region)
+	}
```
Comment (rfranzke) [inline pkg/controller/worker/machine_images.go:78]: Use `pointer.StringDeref` and remove this `if/else` construct

File: pkg/controller/worker/machine_images.go [hunk 7]
```diff
@@ -61,13 +61,21 @@ func (w *workerDelegate) findMachineImage(name, version string, region string, a
 
 		machineImage, err := helper.FindMachineImage(workerStatus.MachineImages, name, version, arch)
 		if err != nil {
-			return "", worker.ErrorMachineImageNotFound(name, version, *arch, region)
+			if arch != nil {
+				return "", worker.ErrorMachineImageNotFound(name, version, *arch, region)
+			} else {
+				return "", worker.ErrorMachineImageNotFound(name, version, region)
+			}
```
Comment (rfranzke) [inline pkg/controller/worker/machine_images.go:68]: same here

File: pkg/apis/aws/helper/helper.go (hunk 3 above)
Comment (rfranzke) [reply pkg/apis/aws/helper/helper.go]: I don't see this being addressed, or how does it work?

File: docs/usage-as-operator.md (hunk 5 above)
Comment (acumino) [reply docs/usage-as-operator.md:31]: PTAL c6870a2f6d57108c294fb0b0aec6a4ff2e075e0e.

File: pkg/apis/aws/helper/helper.go (hunk 3 above)
Comment (acumino) [reply pkg/apis/aws/helper/helper.go]: Done in this c6870a2f6d57108c294fb0b0aec6a4ff2e075e0e

File: pkg/controller/worker/machines.go [hunk 8]
```diff
@@ -111,14 +112,17 @@ func (w *workerDelegate) generateMachineConfig() error {
 			return err
 		}
 
-		ami, err := w.findMachineImage(pool.MachineImage.Name, pool.MachineImage.Version, w.worker.Spec.Region)
+		arch := pointer.StringDeref(pool.Architecture, v1beta1constants.ArchitectureAMD64)
```
Comment (rfranzke) [inline pkg/controller/worker/machines.go:115]: Nice, I like this :) Last comment: Can you add a test case for this, i.e. have a worker pool without the `architecture`? Maybe you can just remove https://github.com/gardener/gardener-extension-provider-aws/pull/565/files#diff-12a1fa4ba4f4250731a53989c476c9407e832ccecae17ad3fc1b2eee100ad7c3R431?

Comment (acumino) [reply pkg/controller/worker/machines.go:115]: Yes removing this line implicitly test the required behavior.

**Review by rfranzke — APPROVED** [review]
Thanks!

/lgtm

**Pull request merged**